### PR TITLE
Implement inbound connection banning support (AGORA side)

### LINKED ISSUE #202: Implement inbound connection banning support in vibe.d
As a Developer, I need to prevent one form of DDoS attacks from other nodes, we need support for banning inbound connections from specific addresses.

**Definition of Done:**
- [x] When a request reaches the server, it can be made to be rejected before any significant processing / memory allocation happens
- [x] The functionality supports dynamically banning
- [x] The functionality supports dynamically unbanning
- [x] The functionality does not add much overhead on the request processing

### PR DESCRIPTION
**Pull request by ferencdg**
fixes: #202

### CODE REVIEW

File: source/agora/utils/InetUtils.d [hunk 1]
```diff
@@ -19,12 +19,24 @@ import std.algorithm;
...
 
 mixin AddLogger!();
 
+///
+public enum HostType : ubyte
+{
+    IPv4,
+    IPv6,
+    Domain,
+    Unknown
```
Comment (Geod24) [inline source/agora/utils/InetUtils.d]:
```suggestion
    Unknown,
```

File: source/agora/utils/InetUtils.d [hunk 2]
```diff
@@ -174,6 +186,135 @@ version (Windows)
     }
 }
 
+    /***************************************************************************
+        Converts a compressed IPv6 representation to uncompressed representation
```
Comment (Geod24) [inline source/agora/utils/InetUtils.d]:
```suggestion
    /***************************************************************************
    
        Expand an IPv6 address to its canonical representation
```

File: source/agora/utils/InetUtils.d [hunk 3]
```diff
@@ -174,6 +186,135 @@ version (Windows)
...
+
+        Params:
+            ip = compressed IPv6 address
+
+        Returns:
+            uncompressed representation of IPv6 address
+
+    ***************************************************************************/
+
+    public static string convertCompressedIPv6ToUncompressed (string ip) @safe pure
```
Comment (Geod24) [inline source/agora/utils/InetUtils.d]: `expandIPv6` or something shorter please.

Comment (Geod24) [inline source/agora/utils/InetUtils.d]:
```suggestion
    public static string convertCompressedIPv6ToUncompressed (const(char)[] ip) @safe pure
```

AFAICS

File: source/agora/utils/InetUtils.d [hunk 4]
```diff
@@ -174,6 +186,135 @@ version (Windows)
...
+        auto parts = ip.split(":");
+        const first_empty_ind = parts.countUntil("");
+
+        // filter out all the empty parts, except the first one
+        auto res = zip(iota(parts.length), parts).filter!((tup) => tup[0] == first_empty_ind || !tup[1].empty()).
+                   map!(tup => tup[1]).array();
+
+        // fill in the empty part with the needed zeros
+        return res.map!(part => (part.empty() ? replicate(["0"],8 - res.length + 1) : [part]))
+                    .joiner.array.join(":");
```
Comment (Geod24) [inline source/agora/utils/InetUtils.d]:
```suggestion
                    .joiner(':').array();
```

?

File: source/agora/utils/InetUtils.d [hunk 5]
```diff
@@ -174,6 +186,135 @@ version (Windows)
...
+            http://[1:2::3]/data
+
+        Params:
+          url = the URL from which we would like to extract the host and port
+
+        Returns: `HostPortTup` containing the extracted host, port and host type
+
+    ****************************************************************************/
+
+    public static HostPortTup extractHostAndPort (string url) @safe
```
Comment (Geod24) [inline source/agora/utils/InetUtils.d:258]: Couldn't we just use Vibe.d for this ?

**Review by Geod24 — COMMENTED** [review]
Needs a rebase and adaptation to the new names

File: source/agora/network/NetworkManager.d [hunk 6]
```diff
@@ -408,6 +408,13 @@ public class NetworkManager
         }
     }
 
+    /// Returns an already instantiated version of the BanManager
+    /// (please also see `NetworkManager.getBanMananger()`)
+    public BanManager getAlreadyCreatedBanManager () @safe @nogc nothrow pure
```
Comment (AndrejMitrovic) [inline source/agora/network/NetworkManager.d:413]: We might want to rename all the `get*` functions to `make*/create*` functions so we would have:

```
makeBanManager() or createBanManager()
getBanManager()
```

We can leave that for a future PR.

File: source/agora/utils/InetUtils.d (hunk 4 above)
Comment (ferencdg) [reply source/agora/utils/InetUtils.d]: what I would like here is to flatten embedded maps:
```D
    auto a = [["1"],["2"],["0","0"],["5"]];
    writeln(a.map!(e => e).joiner.array.join(":"));
    // should print out 1:2:0:0:5
```

this line
```D
writeln(a.map!(e => e).joiner(":").array; 
```
returns a compile error

File: source/agora/utils/InetUtils.d (hunk 5 above)
Comment (ferencdg) [reply source/agora/utils/InetUtils.d:258]: I might be able to make it work, but it lacks the main thing we need: classifying whether the 'host' is ipv4/ipv6/domain

File: tests/runner.d [hunk 7]
```diff
@@ -34,7 +34,7 @@ private int main (string[] args)
     const lflags = getLflags();
     if (!importPaths.length || !lflags.length)
         return 1;
-    const Args = ["rdmd", "-vcolumns"] ~ args[1 .. $] ~
+    const Args = ["rdmd", "-vcolumns", "-i=stdx"] ~ args[1 .. $] ~
```
Comment (ferencdg) [inline tests/runner.d:37]: vibe.d(through the URL class) seems to pulled in stdx:allocator, let me know if you have a better idea of handling this

Comment (Geod24) [reply tests/runner.d:37]: This should be returned by `getImportPaths`, but I'm fine if we leave this as is.

File: source/agora/utils/InetUtils.d [hunk 8]
```diff
@@ -174,6 +190,145 @@ version (Windows)
     }
 }
 
+    /***************************************************************************
+        Expand an IPv6 address to its canonical representation
```
Comment (Geod24) [inline source/agora/utils/InetUtils.d]: Space

File: source/agora/common/BanManager.d [hunk 9]
```diff
@@ -174,6 +176,23 @@ public class BanManager
...
+                    foreach (ref addr_info; getAddressInfo(host_and_port.host))
+                    {
+                       auto resolved_address = addr_info.address.toAddrString();
+                       if (addr_info.address.addressFamily == AddressFamily.INET6)
+                           resolved_address = InetUtils.expandIPv6(resolved_address);
+
+                       this.ban(resolved_address);
+                    }
+            }
+            catch(Exception e)
```
Comment (Geod24) [inline source/agora/common/BanManager.d]: Space

File: source/agora/utils/InetUtils.d [hunk 10]
```diff
@@ -322,10 +322,6 @@ version (Windows)
 
         assert(extractHostAndPort("seed.bosagora.io/blabla")
             == HostPortTup("seed.bosagora.io", 80, HostType.Domain));
-
-        // Negative tests
-        assert(extractHostAndPort("blab/blabla")
-            == HostPortTup("", 0, HostType.Invalid));
```
Comment (Geod24) [inline source/agora/utils/InetUtils.d]: Why is this removed ?

**Review by Geod24 — APPROVED** [review]
LGTM

File: source/agora/utils/InetUtils.d (hunk 8 above)
Comment (ferencdg) [reply source/agora/utils/InetUtils.d]: for the 'getImportPaths': I was wondering if the reason for missing stdx is that in dub.json the submodule/stdx-allocator is not referenced.

File: source/agora/utils/InetUtils.d (hunk 10 above)
Comment (ferencdg) [reply source/agora/utils/InetUtils.d]: In the newest version of the implementation "blab/blabla" is valid: the implementation will try to detect if the url has any schema or not, and if not then "http" is assumed. 

I can add more negative tests though, if you like.

Comment (Geod24) [reply source/agora/utils/InetUtils.d]: Well then this part of the diff should be squashed in the previous commit. Because the previous commit introduce a test, and this commit removes that diff. Since this commit does not touch the function that is tested, either the previous commit does not pass tests, or the test shouldn't be removed. From what you're telling me it sounds like the former ?

Comment (ferencdg) [reply source/agora/utils/InetUtils.d]: sorry my bad, fixed all the issues by now

File: source/agora/utils/InetUtils.d (hunk 8 above)
Comment (Geod24) [reply source/agora/utils/InetUtils.d]: Still missing the newline
